xarray/backends/rasterio_.py

```diff
@@ -0,0 +1,120 @@
+import numpy as np
+
+try:
+    import rasterio
+except ImportError:
+    rasterio = False
+
+from .. import Variable, DataArray
+from ..core.utils import FrozenOrderedDict, Frozen, NDArrayMixin
+from ..core import indexing
+from ..core.pycompat import OrderedDict, suppress
+
+from .common import AbstractDataStore
+
+__rio_varname__ = 'raster'
+
+
+class RasterioArrayWrapper(NDArrayMixin):
+    def __init__(self, ds):
+        self._ds = ds
+        self.array = ds.read()
+
+    @property
+    def dtype(self):
+        return np.dtype(self._ds.dtypes[0])
+
+    def __getitem__(self, key):
+        if key == () and self.ndim == 0:
+            return self.array.get_value()
+        return self.array[key]
```

Comment:
Based on what @fmaussion said above, I think this should be something like:

```Python
    def __getitem__(self, key):
        if key == () and self.ndim == 0:
            return self._ds.read()
        return self._ds.read(band, window=window)
```

Where `band` and `window` describe the slice of data to be read. Off the top of my head, I'm not exactly sure how to parse the key here though.